internal/states/encryption/encryptionconfig/config.go

```diff
@@ -0,0 +1,213 @@
+	KeyProviderDirect     KeyProviderName = "direct"     // key is explicitly specified in config field "key"
+)
+
+type KeyProviderConfig struct {
+	// Name specifies which key provider to use.
+	Name KeyProviderName `json:"name"`
+
+	// Config configures the key provider.
+	//
+	// The available values are key provider dependent.
+	Config map[string]string `json:"config"`
+}
+
+// NameValid checks that the name has been registered correctly.
+//
+// This is an early check, unlike Validate().
+func (k KeyProviderConfig) NameValid() error {
+	validator, ok := getKeyProviderConfigValidation(k.Name)
+	if !ok || validator == nil {
+		return fmt.Errorf("error in configuration for key provider %s: no registered key provider with this name", k.Name)
+	}
+	return nil
+}
+
+// Validate checks the configuration after it has been merged from all sources.
+func (k KeyProviderConfig) Validate() error {
+	if err := k.NameValid(); err != nil {
+		return err
+	}
+
+	validator, _ := getKeyProviderConfigValidation(k.Name)
+	if err := validator(k); err != nil {
+		return fmt.Errorf("error in configuration for key provider %s: %s", k.Name, err.Error())
+	}
+
+	return nil
+}
+
+type EncryptionMethodName string
+
+const (
+	EncryptionMethodFull EncryptionMethodName = "full" // full state encryption
+)
+
+type EncryptionMethodConfig struct {
+	// Name specifies which encryption method to use.
+	Name EncryptionMethodName `json:"name"`
+
+	// Config configures the key provider.
+	//
+	// The available values are key provider dependent.
+	Config map[string]string `json:"config"`
+}
+
+// NameValid checks that the name has been registered correctly.
+//
+// This is an early check, unlike Validate().
+func (m EncryptionMethodConfig) NameValid() error {
+	validator, ok := getEncryptionMethodConfigValidation(m.Name)
+	if !ok || validator == nil {
+		return fmt.Errorf("error in configuration for encryption method %s: no registered encryption method with this name", m.Name)
+	}
+	return nil
+}
+
+// Validate checks the configuration after it has been merged from all sources.
+func (m EncryptionMethodConfig) Validate() error {
+	if err := m.NameValid(); err != nil {
+		return err
+	}
+
+	validator, _ := getEncryptionMethodConfigValidation(m.Name)
+
+	if err := validator(m); err != nil {
+		return fmt.Errorf("error in configuration for encryption method %s: %s", m.Name, err.Error())
+	}
+
+	return nil
+}
+
+func parseJsonStructure(jsonValue string) (ConfigEnvJsonStructure, error) {
+	parsed := make(ConfigEnvJsonStructure)
+
+	decoder := json.NewDecoder(strings.NewReader(jsonValue))
+	decoder.DisallowUnknownFields()
+	err := decoder.Decode(&parsed)
+	if err != nil {
+		logging.HCLogger().Trace("json parse error", "details", err.Error())
+		return nil, errors.New("json parse error, wrong structure, or unknown fields - details omitted for security reasons (may contain key related settings)")
```

Comment:
Removing this because the error may contain sensitive values and anything above trace would be irresponsible. The structure is fairly simple, so users should be able to work with a better error message (which we are adding).